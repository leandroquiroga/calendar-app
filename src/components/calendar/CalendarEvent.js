import React from 'react'

export const CalendarEvent = ({ event }) => {

  const { title, user } = event;
  return (
    <article className='d-flex flex-column justify-content-center'>
      <span> {title} </span>
      <strong>- { user.name } </strong>
    </article>
  )
}
