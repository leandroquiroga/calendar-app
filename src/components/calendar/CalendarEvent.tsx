import React from 'react'
import { EventProps } from 'react-big-calendar';
import { RootState } from '../../store/store';
import { useSelector } from 'react-redux';

export interface Event {
  title: string;
  notes: string;
  end: string;
  start: string;
  id?: string;
  user: User;
}
type User = {
  uid: string;
  name: string;
};

export const CalendarEvent = ({ event }: EventProps<Event>) => {
  const name = useSelector((state: RootState) => state.auth.name);
  return (
    <article className="d-flex flex-column justify-content-center align-content-center flex-wrap">
      <span className="calendar_event__title"> {event.title} </span>
      <span className="calendar_event__name"> -{name} </span>
      {event.notes && <small className="calendar_event__notes">{event.notes}</small>}
    </article>
  );
};