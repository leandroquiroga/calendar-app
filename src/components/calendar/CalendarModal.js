import React, { useEffect, useState } from 'react'
import { customStylesModal } from '../../helpers/customStylesModal';

import DateTimePicker from 'react-datetime-picker';
import Modal from 'react-modal';
import moment from 'moment';
import { useDispatch, useSelector } from 'react-redux';
import { uiCloseModal } from '../../actions/actions';
import { clearEventActive, eventStartAddNew, eventStartLoading, eventStartUpdate } from '../../actions/events';

// Obtenemos el id #root para el modal

(process.env.NODE_ENV !== 'test') && Modal.setAppElement('#root');

// Valor inicial de la fecha
const nowDateStart = moment().minutes(0).seconds(0).add(1, 'hours');

// Valor final de la fecha
const nowDateEnd = nowDateStart.clone().add(1, 'hours');

// Valor inicial del formulario 
const initalValue = {
  title: '',
  notes: '',
  end: nowDateEnd.toDate(),
  start: nowDateStart.toDate(),
}
export const CalendarModal = () => {
  
  const dispatch = useDispatch();
  const { modalOpen } = useSelector(state => state.ui);
  const { activeEvent } = useSelector(state => state.calendar);

  // Mantiene el estado de la fecha inicial seleccionada 
  const [startDate, setStartDate] = useState(nowDateStart.toDate());

  const [errorMsgDate, setErrorMsgDate] = useState(false)
  
  // Mantiene el estado de la fecha final seleccionada
  const [endDate, setEndDate] = useState(nowDateEnd.toDate());

  const [errorTitle, setErrorTitle] = useState(true);
  
  // Propiedades iniciales del formulario 
  const [formValues, setFormValues] = useState(initalValue);
  
  const { title, notes, start, end } = formValues;
  

  // Al abrir el modal,
  // si exite el activeEvent coloca los valores al formulario
  // sino reinicia al estado inicial el formulario
  useEffect(() => {
    if (activeEvent) {
      setFormValues(activeEvent);
    } else {
      setFormValues(initalValue)
    }
  }, [activeEvent]);

  useEffect(() => {
    dispatch(eventStartLoading());
  }, [dispatch]);

  // Dispara la accion para cerrar el modal, limpia el formulario
  // y reinicia el esta de propiedad activeEvent
  const closeModal = () => {
    dispatch(uiCloseModal());
    setFormValues(initalValue);
    dispatch(clearEventActive())
  }
  
  const handleInputChange = ({ target }) => {
    setFormValues({
      ...formValues,
      [target.name]: target.value,
    })
  };


  const handleSubmitNewEvent = (e) => {
    e.preventDefault();

    // Variables de las fechas inicio y fin
    const momentStart = moment(start);
    const momentEnd = moment(end);


    // Valida que la hora de finalizacion del evento no sea
    // menor al de la inicializacion
    if (momentStart.isSameOrAfter(momentEnd)) { 
     return setErrorMsgDate(true)
    }

    setErrorTitle(false);

    // Corregir este choclo ma√±ana Urgente ! 
    if (title.trim().length < 2) {
      return setErrorTitle(false);
    };

    if (activeEvent) {
      //Actualiza el evento 
      dispatch(eventStartUpdate(formValues))
    } else {
      // Realiza la grabacion del evento
      dispatch(eventStartAddNew(formValues));
    };
    setErrorTitle(true);
    closeModal()
  }
  
  // Cambia el valor de la fecha inicial seleccionada y del campo start
  const handleStartDateChange = (e) => {
    setStartDate(e);
    setFormValues({
      ...formValues,
      start: e,
    });
  };

  // Cambia el valor de la fecha final seleccionada y del campo end
  const handleEndDateChange = (e) => {
    setEndDate(e);
    setFormValues({
      ...formValues,
      end: e,
    })
  };

  return (
    <article className='d-flex justify-content-center'>
      <Modal
        className='modal_container'
        ariaHideApp={process.env.NODE_ENV !== 'test' }
        closeTimeoutMS={200}
        isOpen={modalOpen}
        //onAfterOpen={afterOpenModal}
        onRequestClose={closeModal}
        overlayClassName='modal_container__fondo'
        style={customStylesModal}
      > 
        <h1 className='font-bold text-center fs-5'>
          {(activeEvent) ? 'Editar Evento' : 'Nueva Evento'}
        </h1>
        <hr />
        <form
          className='container'
          onSubmit={handleSubmitNewEvent}
        >
          
          <div className='form-group my-1'>
            <label
              className='form-label'
            >
              Fecha y hora inicio
            </label>
            <DateTimePicker
              className='form_modal__field'
              onChange={handleStartDateChange}
              placeholder='Fecha inicio...'
              value={startDate}
            />
          </div>

          <div className='form-group my-1'>
            <label
              className='form-label'
            >
              Fecha y hora fin
            </label>
            <DateTimePicker
              className='form_modal__field'
              minDate={startDate}
              onChange={handleEndDateChange}
              placeholder='Fecha fin...'
              value={endDate}
            />
          </div>
          {errorMsgDate && <small className='text-center form-text text-danger'>La hora es incorrecta</small> }
          <hr />

          <div className='form-group my-1'>
            <label
              className='form-label'
              htmlFor='inputTitle'
            >
              Titulo y notas
            </label>
            <input 
              autoComplete='off'
              className={`form_modal__field ${ !errorTitle && 'form_modal__field-error' }`}
              name='title'
              onChange={handleInputChange}
              placeholder='Titulo del evento'
              type='text'
              value={title}
            />
            {(errorTitle)
              ? <small className='form-text text-muted'> Una descripcion corta</small>
              : <small className='form-text text-danger'> Debe colocar un titulo </small>
            }
          </div>

          <div className='form-group my-1'>
            <textarea
              className='form_modal__field'
              name='notes'
              onChange={handleInputChange}
              placeholder='Escriba su nota'
              rows='5'
              type='text'
              value={notes}
            ></textarea>
             <small className='form-text text-muted'> Una descripcion corta (opcional)</small>
          </div>

          <button
            className="btn btn-danger btn-block form_modal__button"
            type="submit"
          >
            <i className="far fa-save"></i>
            <span> Guardar</span>
          </button>

        </form>

      </Modal>

    </article>
  )
}