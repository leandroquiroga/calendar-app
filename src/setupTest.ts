import '@testing-library/jest-dom';
import Enzyme from 'enzyme';
import ReactAdapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer,  } from 'enzyme-to-json';

Enzyme.configure({ adapter: new ReactAdapter() });

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}) as any)