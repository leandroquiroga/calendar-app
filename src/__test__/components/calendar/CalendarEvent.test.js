import { mount } from "enzyme";
import { CalendarEvent } from "../../../components/calendar/CalendarEvent";

import '@testing-library/jest-dom';
import '../../../setupTest';


describe('Test on component <CalendarEvent />', () => {
  const event = {
    title:"Testing",
    notes:"This test should run correctly",
    start:"2022-06-02T12:45:00.000Z",
    end: "2022-06-02T13:00:00.000Z",
    user: {
      _id:"6285455405b1fddeb0b5a999",
      name:"Leandro"

    },
    id:"62978ddd74ecd43d8c87ea48",
  };
  
  test('this test should make a snapshot correctly', () => { 
    const wapper = mount(<CalendarEvent event={event} />); 
    expect(wapper).toMatchSnapshot();
  });
});