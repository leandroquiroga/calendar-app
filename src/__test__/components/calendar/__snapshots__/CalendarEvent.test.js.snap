// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on component <CalendarEvent /> this test should make a snapshot correctly 1`] = `
<article
  className="d-flex flex-column justify-content-center align-content-center flex-wrap"
>
  <span
    className="calendar_event__title"
  >
     
    Testing
     
  </span>
  <span
    className="calendar_event__name"
  >
     -
    Leandro
     
  </span>
  <small
    className="calendar_event__notes"
  >
    This test should run correctly
  </small>
</article>
`;
